taichi/codegen/llvm/codegen_llvm.cpp

```diff
@@ -369,47 +420,102 @@ void TaskCodeGenLLVM::visit(UnaryOpStmt *stmt) {
-      } else if (is_integral(from) && is_real(to)) {
-        cast_op = is_signed(from) ? llvm::Instruction::CastOps::SIToFP
-                                  : llvm::Instruction::CastOps::UIToFP;
+    } else if (is_real(from) != is_real(to) ||
+               is_real_tensor(from) != is_real_tensor(to)) {
+      if ((is_real(from) || is_real_tensor(from)) &&
+          (is_integral(to) || is_integral_tensor(to))) {
+        cast_op = (is_signed_tensor(to) || is_signed(to))
+                      ? llvm::Instruction::CastOps::FPToSI
+                      : llvm::Instruction::CastOps::FPToUI;
+      } else if ((is_integral(from) || is_integral_tensor(from)) &&
+                 (is_real(to) || is_real_tensor(to))) {
+        cast_op = (is_signed_tensor(from) || is_signed(from))
+                      ? llvm::Instruction::CastOps::SIToFP
+                      : llvm::Instruction::CastOps::UIToFP;
       } else {
         TI_P(data_type_name(from));
         TI_P(data_type_name(to));
         TI_NOT_IMPLEMENTED;
       }
-      auto cast_type = to->is_primitive(PrimitiveTypeID::f16)
-                           ? PrimitiveType::f32
-                           : stmt->cast_type;
-
-      llvm_val[stmt] = builder->CreateCast(cast_op, llvm_val[stmt->operand],
-                                           tlctx->get_data_type(cast_type));
-
-      if (to->is_primitive(PrimitiveTypeID::f16)) {
-        llvm_val[stmt] = builder->CreateFPTrunc(
-            llvm_val[stmt], llvm::Type::getHalfTy(*llvm_context));
+      bool use_f16 = to->is_primitive(PrimitiveTypeID::f16) ||
+                     (to->is<TensorType>() &&
+                      to->cast<TensorType>()->get_element_type()->is_primitive(
+                          PrimitiveTypeID::f16));
+      auto cast_type = use_f16 ? (to->is<TensorType>()
+                                      ? TypeFactory::create_tensor_type(
+                                            to->cast<TensorType>()->get_shape(),
+                                            PrimitiveType::f32)
+                                      : PrimitiveType::f32)
+                               : stmt->cast_type;
+
+      auto cast_func = [this, cast_op](llvm::Value *value, llvm::Type *type) {
+        return this->builder->CreateCast(cast_op, value, type);
+      };
+      create_value_cast(stmt, cast_func, cast_type);
+
+      if (use_f16) {
+        auto trunc_func = [this](llvm::Value *value, llvm::Type *type) {
+          return this->builder->CreateFPTrunc(value, type);
+        };
+        create_fp_trunc(stmt, trunc_func, llvm::Type::getHalfTy(*llvm_context),
+                        cast_type->is<TensorType>(), /*trunc_self=*/true);
       }
-    } else if (is_real(from) && is_real(to)) {
-      if (data_type_size(from) < data_type_size(to)) {
-        llvm_val[stmt] = builder->CreateFPExt(
-            llvm_val[stmt->operand], tlctx->get_data_type(stmt->cast_type));
+    } else if ((is_real(from) || is_real_tensor(from)) &&
```

Comment:
You can simply check for `is_real(from.get_element_type())`, cause seriously there's no special treatment for `TensorType` here.

Same for most of the checks in `visit(BinaryOpStmt*)`